### PR TITLE
[sk] Added IO config handler

### PR DESCRIPTION
**Pull request by skunichetty**
# Summary
This code enables the user to define data IO configuration settings via a configuration file. In every repo created with `mage 
init`, a new file named `io_config.yaml` will be created, in which all IO configurations can be stored (alternatively from manually specifying the arguments in code). This is more useful, for example, when storing secret keys and access tokens where the secrets can be handled separately from code execution.

Here is a sample config file (contains only the default profile with credentials only for AWS data sources):
```yaml
default:
  AWS:
    Redshift:
      database: your_redshift_database_name
      # Use the settings below if you are using temporary credentials
      user: your_authentication_user_name
      password: your_authentication_password
      host: your_redshift_cluster_hostname
      port: your_redshift_cluster_port
      # Use the settings below if you are using IAM credentials
      cluster_identifier: your_redshift_identifier
      db_user: your_redshift_username
      profile: your_aws_credentials_profile
      iam: True
    S3:
      bucket_name: your_bucket_name
      object_key: your_object_key
    # Omit these settings if aws credentials are already configured on file
    aws_access_key_id: your_aws_access_key_id
    aws_secret_access_key: your_secret_access_key_id
    region_name: your_aws_region
```

To use this configuration file, use the following example code below:
```python
with BaseLoader.with_config(IOConfig(CONFIG_PATH).use('default')) as loader:
     pass
```
which initializes the data loader with the credentials stored in the `default` profile in `io_config.yaml`. The path to the config file should be specified as input to the constructor of the `IOConfig` object. The `.use()` function allows the user to specify the profile they wish to use.

In the example above, the 'AWS' and 'Redshift' objects are read to initialize the `Redshift` data loader. Specifically, the shared credentials from 'AWS' object and the specific 'Redshift' credentials are combined as inputs to the `Redshift` data loader. The implementation for how parameters are read from the configuration file are data loader dependent.

# Tests
Testing was performed to make sure that all data loaders still worked while loading this configuration file.

cc:
@wangxiaoyou1993

### CODE REVIEW

**Comment by skunichetty** [comment]
Don't review just yet, I need to rebase with master to update with data exporter changes.

File: mage_ai/data_preparation/templates/data_exporters/bigquery.py [hunk 1]
```diff
@@ -25,13 +26,14 @@ def export_data_to_big_query(df: DataFrame) -> None:
...
+
+    Alternatively, all parameters can be specified in the configuration file.
     """
     table_id = 'your-project.your_dataset.your_table_name'
-    config = {
-        # Specify any other configuration settings here to pass to BigQuery client
-        'project': 'your_project_name',
-    }
-    return BigQuery(**config).export(
+    config_path = 'path/to/your/io/config/file.yaml'
```
Comment (wangxiaoyou1993) [inline mage_ai/data_preparation/templates/data_exporters/bigquery.py:33]: Let's set a default path here since we auto generated the io_config.yaml

Comment (skunichetty) [reply mage_ai/data_preparation/templates/data_exporters/bigquery.py]: Updated default path to use the autogenerated config path in templates.

File: mage_ai/data_preparation/templates/data_exporters/bigquery.py [hunk 2]
```diff
@@ -25,13 +26,14 @@ def export_data_to_big_query(df: DataFrame) -> None:
...
+
+    Alternatively, all parameters can be specified in the configuration file.
     """
     table_id = 'your-project.your_dataset.your_table_name'
-    config = {
-        # Specify any other configuration settings here to pass to BigQuery client
-        'project': 'your_project_name',
-    }
-    return BigQuery(**config).export(
+    config_path = './default_repo/io_config.yaml'
```
Comment (wangxiaoyou1993) [inline mage_ai/data_preparation/templates/data_exporters/bigquery.py:33]: let's use `get_repo_path` from `mage_ai.data_preparation.repo_manager` to get the correct repo path.

File: mage_ai/io/bigquery.py [hunk 3]
```diff
@@ -99,7 +99,11 @@ def export(
         self.client.load_table_from_dataframe(df, table_id, job_config=config).result()
 
     @classmethod
-    def with_credentials_file(cls, path_to_credentials: str, **kwargs):
+    def with_config(cls, config: Mapping[str, Any]) -> 'BigQuery':
+        return cls(**config['BigQuery'])
```
Comment (wangxiaoyou1993) [inline mage_ai/io/bigquery.py:103]: Let's define those config keys as constants and raise exception with good error message when config is invalid.

**Review by wangxiaoyou1993 — APPROVED** [review]

File: mage_ai/data_preparation/templates/data_exporters/bigquery.py (hunk 2 above)
Comment (skunichetty) [reply mage_ai/data_preparation/templates/data_exporters/bigquery.py]: Fixed, templates now use `get_repo_path`

File: mage_ai/io/bigquery.py (hunk 3 above)
Comment (skunichetty) [reply mage_ai/io/bigquery.py]: Moved config keys into an enum `IOConfigKeys`.

**Pull request merged**